sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StructuredMessageDecoder.java

```diff
@@ -0,0 +1,182 @@
+        int toRead = Math.min(buffer.remaining(), currentSegmentContentLength - currentSegmentContentOffset);
+        toRead = Math.min(toRead, size);
+
+        if (toRead == 0)
+            return;
+
+        byte[] content = new byte[toRead];
+        buffer.get(content);
+        output.write(content, 0, toRead);
+
+        if (flags == StructuredMessageFlags.STORAGE_CRC64) {
+            segmentCrc64 = StorageCrc64Calculator.compute(content, segmentCrc64);
+            messageCrc64 = StorageCrc64Calculator.compute(content, messageCrc64);
+        }
+
+        messageOffset += toRead;
+        currentSegmentContentOffset += toRead;
+
+        // Validate the actual bytes read vs expected segment size
+        if (currentSegmentContentOffset > currentSegmentContentLength) {
+            throw new IllegalArgumentException("Segment size mismatch detected in segment " + currentSegmentNumber);
+        }
+
+        if (currentSegmentContentOffset == currentSegmentContentLength) {
+            readSegmentFooter();
+        }
+    }
+
+    private void readSegmentFooter() {
+        if (currentSegmentContentOffset != currentSegmentContentLength) {
+            throw new IllegalArgumentException("Segment content length mismatch in segment " + currentSegmentNumber
+                + ". Expected: " + currentSegmentContentLength + ", Read: " + currentSegmentContentOffset);
+        }
+
+        if (flags == StructuredMessageFlags.STORAGE_CRC64) {
+            if (buffer.remaining() < CRC64_LENGTH) {
+                throw new IllegalArgumentException("Segment footer is incomplete.");
+            }
+
+            long reportedCrc64 = buffer.getLong();
+            if (segmentCrc64 != reportedCrc64) {
+                throw new IllegalArgumentException("CRC64 mismatch detected in segment " + currentSegmentNumber);
+            }
+            segmentCrcs.put(currentSegmentNumber, segmentCrc64);
+            messageOffset += CRC64_LENGTH;
+        }
+
+        if (currentSegmentNumber == numSegments) {
+            readMessageFooter();
+        } else {
+            readSegmentHeader();
+        }
+    }
+
+    private void readMessageFooter() {
+        if (flags == StructuredMessageFlags.STORAGE_CRC64) {
+            if (buffer.remaining() < CRC64_LENGTH) {
+                throw new IllegalArgumentException("Message footer is incomplete.");
+            }
+
+            long reportedCrc = buffer.getLong();
+            if (messageCrc64 != reportedCrc) {
+                throw new IllegalArgumentException("CRC64 mismatch detected in message footer.");
+            }
+            messageOffset += CRC64_LENGTH;
+        }
+
+        if (messageOffset != messageLength) {
+            throw new IllegalArgumentException("Decoded message length does not match expected length.");
+        }
+    }
+
+    public byte[] decode(int size) {
+        ByteArrayOutputStream decodedContent = new ByteArrayOutputStream();
```

Comment:
One more public method we are going to need is a finalize method (I'm not too tied to that name if you have a better one) that will tell the decoder that we have finished decoding and have it validate that we've consumed the entire message. I'm suggesting this because I'm not sure if the case is handled where the structured message get cut off and we only process a portion of the structured message. If we call this finalize after feeding through all byte buffers, it would then throw an error if in fact we did not get complete the decoding process.